ftdetect/lean.lua

```diff
@@ -0,0 +1,65 @@
+_LEAN3_STANDARD_LIBRARY = '.*/[^/]*lean[%-]+3.+/lib/'
+local _LEAN3_VERSION_MARKER = '.*lean_version.*".*:3.*'
+local _LEAN4_VERSION_MARKER = '.*lean_version.*".*lean4:.*'
+
+local find_project_root = require('lspconfig.util').root_pattern(
+  'leanpkg.toml',
+  'lakefile.lean',
+  'lean-toolchain'
+)
+
+lean_nvim_ft_options = {
+  default = "lean",
+  nomodifiable = {
+    '.*/src/lean/.*',       -- Lean 4 standard library
+    '.*/lib/lean/src/.*',   -- Lean 4 legacy standard library
+    '.*/lean_packages/.*',  -- Lean 4 dependencies
+    _LEAN3_STANDARD_LIBRARY .. '.*',
+    '/_target/.*/.*.lean'   -- Lean 3 dependencies
+  }
+}
+
+local function detect(filename)
+  if filename:match('^fugitive://.*') then
+    filename = pcall(vim.fn.FugitiveReal, filename)
+  end
+
+  local abspath = vim.fn.fnamemodify(filename, ":p")
+  local filetype = lean_nvim_ft_options.default
```

Comment:
I think this breaks someone changing the ft default (and probably we don't have a test covering that). In other words, we support changing what to do with single file `.lean` files from lean 4 to lean 3 via a config option, and I think that setting may be being ignored now. But that's just from reading this mentally. Can you double check?

Reply:
This is the config option (from the README):

>   ft = {
    -- What filetype should be associated with standalone Lean files?
    -- Can be set to "lean3" if you prefer that default.
    -- Having a leanpkg.toml or lean-toolchain file should always mean
    -- autodetection works correctly.
    default = "lean",

Reply:
nope! this still works
how i understand it is that if you choose to initialize the plugin you can still change the behaviour before ft detection runs
e.g. the control flow goes something like
1. ftdetect script loaded
2. plugin loaded
3. autocmd for ftdetect runs

I confirmed this by setting `ft = { default = "lean5", }`, and then edited a `test.lean` empty file, then `set ft?` showed `lean5` as the filetype.

The main issue is that if you _don't_ load the plugin (e.g. don't call `require 'lean'.setup` during initialization) then you won't be able to configure the default at all.
We could get around this by making the `lean_nvim_ft_options` global a part of the public API, and extend it in the ftdetect script instead of overwriting it---but that's up to you.